goalc/compiler/compilation/State.cpp

```diff
@@ -27,8 +51,12 @@ Val* Compiler::compile_define_state_hook(const goos::Object& form,
                {"code", {true, {}}},
            });
 
+  // clear previous saved compiled vals
+  m_defstate_code_value = NULL;
```

Comment:
I've tried really hard to avoid having any "global" state like this in the compiler, other than stuff like symbol/typesystem that is actually global.  It can cause some really confusing bugs if compilation aborts/throws an exception (now you have a pointer to some freed `Val`) or if somebody nests `defstate`s (silly, but you could technically do this).  Then the outer defstate would end up using the types from the inner `defstate`.

Maybe we could have `compile_state_handler_set` take a `TypeSpec* code_type_out` and `bool* did_set_out`  argument and set those, like:
```
if (code wasn't #f) {
  // .. 
  if (code_type_out) {
    *code_type_out = val->get_type();
  }
  if (did_set_out) {
    *did_set_out = true;
  }
}

```

Reply:
Oh that's a much better idea.